### ISSUE TITLE
[HostOS devel] [qemu] virsh dompmwakeup fails to wake VM from dompmsuspend state

### ISSUE THREAD

**Issue by balamuruhans** [open-power-host-os/qemu#31]
<cde:info> Mirrored with LTC bug https://bugzilla.linux.ibm.com/show_bug.cgi?id=161564 </cde:info>

virsh dompmwakeup fails to wake VM from dompmsuspend state

**VM state remains running instead of pmsuspended even in suspended state.**

**Steps to reproduce the issue**
1. Start a VM and ensure it reaches login prompt
2. Ensure qemu-guest-agent is available and running in the VM
3. Suspend the VM to mem using virsh dompmsuspend <domain> mem --duration 0
4. VM enters suspended state, try to wake up the VM using virsh dompmwakeup <domain>
5. Libvirt returns that VM got wokeup but ideally it doesn't

```
# virsh list --all
 Id    Name                           State
----------------------------------------------------
 2     virt-tests-vm1                 running

# virsh dompmsuspend virt-tests-vm1 mem --duration 0
Domain virt-tests-vm1 successfully suspended
```
**Guest console after dompmsuspend executed**
```
[root@localhost ~]# [  142.850469] PM: suspend entry (s2idle)
[  142.851306] PM: Syncing filesystems ... done.
[  142.894060] Freezing user space processes ... (elapsed 0.001 seconds) done.
[  142.896827] OOM killer disabled.
[  142.897416] Freezing remaining freezable tasks ... (elapsed 0.001 seconds) done.
[  142.899892] Suspending console(s) (use no_console_suspend to debug)
```
**In Host:**
After suspended the state of the domain remains running but it should be pmsuspended
```
 # virsh list --all
 Id    Name                           State
----------------------------------------------------
 2     virt-tests-vm1                 running

# virsh dompmwakeup virt-tests-vm1
Domain virt-tests-vm1 successfully woken up

```
But actually guest din't woke up.

**Test Environment**

_Libvirt version_
```
# rpm -qa | grep libvirt
libvirt-daemon-driver-nodedev-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-storage-core-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-storage-scsi-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-python-3.2.0-3.el7_4.1.ppc64le
libvirt-libs-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-nwfilter-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-config-nwfilter-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-secret-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-client-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-network-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-lxc-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-storage-logical-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-storage-disk-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-storage-iscsi-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-qemu-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-devel-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-interface-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-config-network-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-storage-mpath-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
libvirt-daemon-driver-storage-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
```

_Qemu version_
```
# rpm -qa | grep qemu
qemu-img-2.10.0-3.dev.gitbf0fd83.el7.centos.ppc64le
qemu-common-2.10.0-3.dev.gitbf0fd83.el7.centos.ppc64le
qemu-system-ppc-2.10.0-3.dev.gitbf0fd83.el7.centos.ppc64le
libvirt-daemon-driver-qemu-3.6.0-3.dev.gitdd9401b.el7.centos.ppc64le
ipxe-roms-qemu-20170123-1.git4e85b27.el7_4.1.noarch
qemu-2.10.0-3.dev.gitbf0fd83.el7.centos.ppc64le
qemu-system-x86-2.10.0-3.dev.gitbf0fd83.el7.centos.ppc64le
```

_Kernel version_
```
# uname -a
Linux bos212lpar 4.14.0-3.dev.git68b4afb.el7.centos.ppc64le #1 SMP Fri Nov 17 08:36:50 UTC 2017 ppc64le ppc64le ppc64le GNU/Linux

```

**Comment by cdeadmin** [comment]
------- Comment From danielhb@br.ibm.com 2017-11-29 14:42:26 EDT-------
I'll post here some findings hoping that anyone more versed in Libvirt than I am weights-in if I got something wrong.

From Libvirt perspective, &quot;dompmsuspend&quot; is a call to the QEMU agent that executes a pm-utils command in the guest (pm-suspend, pm-hibernate ...) depending on the parameter you pass (e.g. &quot;dompmsuspend &lt;domain&gt; mem&quot; will end up calling &quot;pm-suspend&quot; in the guest). If &quot;pm-utils&quot; isn't installed in the guest, the agent will make an effort to emulate the same behavior itself, so it is recommended to have pm-utils but it's not required. Having pm-utils installed in the guest didn't change the behavior - the bug still happens.

One thing to notice is that the qemu agent will *NOT* emit a success response when the pm-suspend command executes. You can see more info in this patch:

commit 432d29db0db9d08fe34a57b8c03af20c9f759d77
Author: Luiz Capitulino &lt;lcapitulino@redhat.com&gt;
Date:   Tue May 8 14:24:48 2012 -0300

    qemu-ga: guest-suspend-ram: don't emit a success response

The docs at qga/qapi-schema.json mentions:

&quot;# This command does NOT return a response on success. There are two options
# to check for success:
#   1. Wait for the SUSPEND QMP event from QEMU
#   2. Issue the query-status QMP command to confirm the VM status is
#      &quot;suspended&quot;
&quot;

Inside QEMU I can see that, as far as Power code goes, there is no one setting the VM state to &quot;suspended&quot; in the workflow. The only place I see the VM being changed to SUSPENDED is inside hw/acpi/core.c , an ACPI control code used by x86 and others (but not Power).  Libvirt returns &quot;Domain successfully suspended&quot; but in reality what happened is that the command to suspend was executed without errors - it doesn't mean the VM state is now SUSPENDED.

And then we go to dompmwakeup. This one is a bit more straightforward: it the machine is in SUSPENDED mode, it changes the runtime of the machine to RUNNING and fire up an event. In our case the machine never left the RUNNING state so dompmwakeup does nothing. So yeah, in the end the VM is suspended but we can't bring it up again.

So, to sum up:

- libvirt reports that the domain was suspended, which isn't true as far as QEMU state goes. But the guest OS is RAM suspended;
- dompmwakeup doesn't try to wake up the guest because the VM state as far as QEMU goes is &quot;RUNNING&quot;.

I'll start by investigating how we can pass the VM to SUSPENDED state when executing dompmsuspended in pseries. I'm also CCing Vasant Hege from RAS team to ask for help

**Comment by cdeadmin** [comment]
------- Comment From danielhb@br.ibm.com 2017-12-06 07:28:42 EDT-------
After more investigation on the code and discussions with Mike Roth, the conclusion is that we're facing here not just a bug but a feature request. The only architecture that fully supports SUSPEND and wake up operations using pm-suspend, in QEMU, is x86. All other architecture haven't got that far yet.

When dompmsuspend is executed with the 'mem' parameter, it triggers guest-suspend-ram in the QEMU guest agent (QGA). This function works by either calling pm-suspend to invoke the SUSPEND or by doing itself. To do itself (and I believe pm-suspend does something similar, but I haven't checked) , QGA writes &quot;mem&quot; at /sys/power/state, which is the Power Management Interface from the kernel. Writing &quot;mem&quot; in this file causes the kernel to execute a Suspend-to-RAM, saving device states (CPU included) in the memory and putting them into suspension.

Speaking briefly about the x86 implementation: when running a QEMU x86 guest, QEMU-side ACPI code linked to the /sys/power/state interface causes the VM to take action on the OS-induced suspension. This includes setting ACPI registers, setting the VM to runstate SUSPENDED and other pre-suspend actions to allow the guest to wake up later on. There are also QEMU-side ACPI code that is activated when we execute &quot;dompmwakeup&quot;, which is simply a call to &quot;system_wakeup&quot; at monitor.

We don't have these triggers implemented in pseries. When pm-suspend is executed, the OS suspends execution but, as far as QEMU is concerned, nothing happened. Since nothing happened, the VM remains at RUNNING state, making system_wakeup do nothing as I described in my previous comment. I've investigated a little how to implement this feature in pseries, definitely possible, but it'll require more thought and investigation on how PAPR suspends/wakes up from an OS induced state.

**Comment by cdeadmin** [comment]
------- Comment From danielhb@br.ibm.com 2017-12-06 07:40:58 EDT-------
All that said, we can't just ignore that dompmsuspend / wakeup does not work as intended in pseries. I've implemented a new QMP/HMP API command &quot;query-wakeup-from-suspend-support&quot; to help us deal with this for the time being.

This new command inspects QEMU internals to tell the user/management whether the running guest has the necessary tools to properly wake up from SUSPENDED state. In case the support isn't available, management can refrain from allowing the suspend operation in the first place.

This is the output of the command when running a pseries guest:

(qemu) info wakeup_from_suspend_support 
wake up from suspend support: disabled
(qemu) 

And running a x86 guest:

(qemu) info wakeup_from_suspend_support 
wake up from suspend support: enabled
(qemu) 

I've also changed the descriptions of both guest-suspend-ram and guest-suspend-hybrid from QEMU guest agent, mentioning that the user/management is now required to not just see if system_wakeup is enabled, but also to see if the wake up support is present.

With this new command and the doc changes, Libvirt can then change the dompmsuspend calls accordingly, avoiding calling guest-suspend-* from the QEMU guest agent if the guest does not have the proper wake up support (everyone but x86 at this moment).

The series are available in the mailing list for review/approval. I'll keep this bugzilla posted with updates.

Daniel

**Comment by cdeadmin** [comment]
------- Comment From bssrikanth@in.ibm.com 2018-05-16 23:59:16 EDT-------
Do we have any updates here?

**Comment by cdeadmin** [comment]
------- Comment From seg@us.ibm.com 2018-05-17 01:10:22 EDT-------
Daniel is probably mostly tied up, at the moment, I suspect. That said, it looks like there wasn't any reply to his v4 patches back in January. It's probably time to poke the maintainers.

**Comment by cdeadmin** [comment]
------- Comment From danielhb@br.ibm.com 2018-05-17 09:16:27 EDT-------
Curiously enough, I received word from one of the maintainers last Monday about these patches. I am tied up with other bugs ATM but I'm managing to send further reviews to get these accepted at last. Let's see how it goes.

**Comment by cdeadmin** [comment]
------- Comment From seg@us.ibm.com 2018-09-14 12:48:09 EDT-------
No longer making plans for future hostos-specific bugs.
